### PR TITLE
Test: Vault Reconciliation

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Add reconciliation checks to all resource vaults

### CODE REVIEW

File: radix-engine-derive/src/scrypto_event.rs [hunk 1]
```diff
@@ -19,6 +19,8 @@ pub fn handle_scrypto_event(input: TokenStream) -> Result<TokenStream> {
     // in the re-interface crate?
     let derive = quote! {
         impl ScryptoEvent for #ident {
+            const EVENT_NAME: &'static str = #ident_string;
+
             fn event_name() -> &'static str {
```
Comment (iamyulong) [inline radix-engine-derive/src/scrypto_event.rs:24]: We only need one of them?

File: radix-engine/src/system/checkers/resource_event_checker.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+#[derive(Debug, Default)]
+pub struct ResourceEventCheckerResults {
+    pub total_supply: BTreeMap<ResourceAddress, Decimal>,
+    pub vault_amounts: BTreeMap<NodeId, Decimal>,
+}
+
+impl ApplicationEventChecker for ResourceEventChecker {
+    type ApplicationEventCheckerResults = ResourceEventCheckerResults;
+
+    fn on_event(
```
Comment (iamyulong) [inline radix-engine/src/system/checkers/resource_event_checker.rs:36]: Thanks for adding recon for both total supply and vault balances!

File: scrypto-unit/src/test_runner.rs [hunk 3]
```diff
@@ -396,16 +384,29 @@ pub struct TestRunner<E: NativeVmExtension, D: TestDatabase> {
...
-        println!("{:?}", results);
+        println!("{:#?}", db_results);
+
+        let event_results = SystemEventChecker::<ResourceEventChecker>::new()
+            .check_all_events(&self.database, &self.collected_events)
+            .expect("Events should be consistent");
+        println!("{:#?}", event_results);
+
+        // If free credits (xrd from thin air) have been used then reconciliation will fail
+        // due to missing mint events
```
Comment (iamyulong) [inline scrypto-unit/src/test_runner.rs:405]: Are we planning to expose `free_credit_used_in_xrd` to receipt, so we can enable recon for all? Happy either way :)

**Review by iamyulong — APPROVED** [review]

File: scrypto-unit/src/test_runner.rs (hunk 3 above)
Comment (talekhinezh) [reply scrypto-unit/src/test_runner.rs:411]: Ran into some complications with `free_credit_used_in_xrd` so avoided this for now :)

**Pull request merged**